WalletProvider.tsx

```diff
@@ -0,0 +1,494 @@
+          address,
+          connection,
+          addressShort: shortenAddress(address),
+          addressDomain: domain,
+          balance
+        }
+      }))
+
+      localStorage.setItem('web3-wallets-name', WALLET_NAMES.Phantom)
+      return true
+    } catch (err) {
+      // @ts-ignore
+      if (err.code === 4001) {
+        console.warn('[Wallet] User rejected the request.')
+      }
+      console.error('[Wallet]', err)
+      return false
+    }
+  }
+
+  const connect = async ({ name, chainId }): Promise<boolean> => {
+    console.log('Wallet.connect()', name, chainId)
+    if (!WALLET_NAMES[name]) {
+      console.error(`Unknown wallet name: ${name}`)
+      return false
+    }
+
+    if (name === 'MetaMask') {
+      if (!window.ethereum || !window.ethereum.isMetaMask) {
+        goMetamask()
+        return false
+      }
+      return connectMetamask(chainId)
+    }
+
+    if (name === 'WalletConnect') {
+      return connectWC(chainId)
+    }
+
+    if (name === 'Phantom') {
+      const isPhantomInstalled = window.solana && window.solana.isPhantom
+      if (!isPhantomInstalled) {
+        goPhantom()
+        return false
+      }
+      return connectPhantom(chainId)
+    }
+
+    return false
+  }
+
+  const restore = async () => {
+    console.log('Wallet.restore()')
+    const walletData = localStorage.getItem('web3-wallets-data')
+
+    if (walletData) {
+      const { name, chainId } = JSON.parse(walletData) as {
+        name: string
+        chainId: string | number
+      }
+
+      return connect({ name, chainId })
+    }
+
+    return false
+  }
+
+  const evmChainChangeHandler = async chainIdHex => {
+    const chainId = parseInt(chainIdHex)
+    console.log('* chainChanged', chainIdHex, chainId)
+    setState(prev => ({
+      ...prev,
+      ...{
+        chainId
+      }
+    }))
+  }
+
+  const evmChangeNetwork = async (provider, params): Promise<boolean> => {
+    const newChainIdHex = params[0].chainId
+
+    try {
+      await provider.send('wallet_switchEthereumChain', [
+        {
+          chainId: newChainIdHex
+        }
+      ])
+      return true
+    } catch (error) {
+      console.warn('Cant change network:', error)
+
+      // the chain has not been added to MetaMask
+      try {
+        console.log('Try to add the network...', params)
+        await provider.send('wallet_addEthereumChain', params)
+        // todo:
+        // Users can allow adding, but not allowing switching
+        return true
+      } catch (addError) {
+        console.warn('Cant add the network:', addError)
+        return false
+      }
+    }
+  }
+
+  const disconnect = () => {
+    console.log('Wallet.disconnect()')
+
+    if (state.name === 'MetaMask' || state.name === 'WalletConnect') {
+      if (state.walletProvider) {
+        state.walletProvider.removeAllListeners('chainChanged', evmChainChangeHandler)
+        state.walletProvider.removeAllListeners('accountsChanged', evmAccountChangeHandler)
```

Comment:
`removeAllListeners` method accepts no parameters and clears all handlers.
 It is not necessary to call it twice 😉
```suggestion
        state.walletProvider.removeAllListeners()
```